File: src/services/recommendationEngine.ts
import { ChargingStation, Connector, EVProfile, RecommendationResult } from '../types';

export const EV_PRESETS: EVProfile[] = [
  // --- Tata Motors ---
  {
    id: 'nexon-ev',
    brand: 'Tata Motors',
    model: 'Nexon EV Long Range',
    batteryKwh: 40.5,
    maxDcKw: 50,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 325
  },
  {
    id: 'curvv-ev',
    brand: 'Tata Motors',
    model: 'Curvv.ev 55',
    batteryKwh: 55.0,
    maxDcKw: 70,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 502
  },
  {
    id: 'punch-ev',
    brand: 'Tata Motors',
    model: 'Punch.ev Empowered+',
    batteryKwh: 35.0,
    maxDcKw: 50,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 365
  },
  {
    id: 'tiago-ev',
    brand: 'Tata Motors',
    model: 'Tiago.ev XZ+',
    batteryKwh: 24.0,
    maxDcKw: 25,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 250
  },

  // --- Mahindra ---
  {
    id: 'xuv400',
    brand: 'Mahindra',
    model: 'XUV400 EL Pro',
    batteryKwh: 39.4,
    maxDcKw: 50,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 375
  },
  {
    id: 'be-6e',
    brand: 'Mahindra',
    model: 'BE 6e INGLO',
    batteryKwh: 59.0,
    maxDcKw: 175,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 550
  },
  {
    id: 'xev-9e',
    brand: 'Mahindra',
    model: 'XEV 9e Flagship',
    batteryKwh: 79.0,
    maxDcKw: 175,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 656
  },

  // --- MG Motor ---
  {
    id: 'mg-windsor',
    brand: 'MG Motor',
    model: 'Windsor EV Essence',
    batteryKwh: 38.0,
    maxDcKw: 45,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 331
  },
  {
    id: 'mg-zs-ev',
    brand: 'MG Motor',
    model: 'ZS EV Exclusive Plus',
    batteryKwh: 50.3,
    maxDcKw: 75,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 461
  },
  {
    id: 'mg-comet',
    brand: 'MG Motor',
    model: 'Comet EV Pace',
    batteryKwh: 17.3,
    maxDcKw: 3.3,
    supportedConnectors: ['Type2'],
    rangeKm: 230
  },

  // --- Hyundai ---
  {
    id: 'ioniq-5',
    brand: 'Hyundai',
    model: 'Ioniq 5 AWD (800V)',
    batteryKwh: 72.6,
    maxDcKw: 350,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 631
  },
  {
    id: 'creta-ev',
    brand: 'Hyundai',
    model: 'Creta Electric',
    batteryKwh: 45.0,
    maxDcKw: 60,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 410
  },
  {
    id: 'kona-ev',
    brand: 'Hyundai',
    model: 'Kona Electric Premium',
    batteryKwh: 39.2,
    maxDcKw: 50,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 452
  },

  // --- Kia ---
  {
    id: 'kia-ev6',
    brand: 'Kia',
    model: 'EV6 GT-Line AWD',
    batteryKwh: 77.4,
    maxDcKw: 350,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 708
  },
  {
    id: 'kia-ev9',
    brand: 'Kia',
    model: 'EV9 Luxury GT',
    batteryKwh: 99.8,
    maxDcKw: 350,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 561
  },

  // --- BYD ---
  {
    id: 'byd-atto3',
    brand: 'BYD',
    model: 'Atto 3 Blade Dynamic',
    batteryKwh: 60.5,
    maxDcKw: 80,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 521
  },
  {
    id: 'byd-seal',
    brand: 'BYD',
    model: 'Seal Performance AWD',
    batteryKwh: 82.5,
    maxDcKw: 150,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 580
  },
  {
    id: 'byd-emax7',
    brand: 'BYD',
    model: 'eMAX 7 Superior MPV',
    batteryKwh: 71.8,
    maxDcKw: 115,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 530
  },

  // --- Citroën ---
  {
    id: 'citroen-ec3',
    brand: 'Citroën',
    model: 'ë-C3 Shine Turbo',
    batteryKwh: 29.2,
    maxDcKw: 30,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 320
  },

  // --- Luxury Tier (Volvo, BMW, Mercedes) ---
  {
    id: 'volvo-ex40',
    brand: 'Volvo',
    model: 'EX40 Recharge Twin',
    batteryKwh: 78.0,
    maxDcKw: 150,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 534
  },
  {
    id: 'bmw-i4',
    brand: 'BMW',
    model: 'i4 eDrive40 Gran Coupé',
    batteryKwh: 83.9,
    maxDcKw: 205,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 590
  },
  {
    id: 'mercedes-eqb',
    brand: 'Mercedes-Benz',
    model: 'EQB 350 4MATIC',
    batteryKwh: 70.5,
    maxDcKw: 100,
    supportedConnectors: ['CCS2', 'Type2'],
    rangeKm: 423
  },

  // --- Two-Wheelers ---
  {
    id: 'ather-450x',
    brand: 'Ather Energy',
    model: '450X Gen 3 Pro',
    batteryKwh: 3.7,
    maxDcKw: 3.3,
    supportedConnectors: ['Type2'],
    rangeKm: 110
  },
  {
    id: 'ola-s1-pro',
    brand: 'Ola Electric',
    model: 'S1 Pro Gen 2',
    batteryKwh: 4.0,
    maxDcKw: 3.3,
    supportedConnectors: ['Type2'],
    rangeKm: 195
  }
];

/**
 * Evaluates stations and selects the single "Best Match" for the active vehicle profile.
 */
export function calculateBestMatch(
  stations: ChargingStation[],
  profile: EVProfile
): RecommendationResult | null {
  if (!stations.length) return null;

  let bestStation: ChargingStation | null = null;
  let topScore = -Infinity;
  let bestConnector: Connector | null = null;
  let matchReason = '';

  for (const station of stations) {
    // 1. Filter compatible connectors
    const compatible = station.connectors.filter(c => profile.supportedConnectors.includes(c.type));
    if (!compatible.length) continue;

    // Separate available vs busy
    const availableConnectors = compatible.filter(c => c.status === 'AVAILABLE');
    const targetPool = availableConnectors.length > 0 ? availableConnectors : compatible;

    // Pick highest power connector in that pool
    const selectedConn = targetPool.reduce((best, cur) => cur.powerKw > best.powerKw ? cur : best, targetPool[0]);

    // 2. Compute Power Suitability
    // If car caps at 50kW, receiving 50-60kW is 100% efficient.
    // Paying extra for 150kW+ when capped at 50kW has diminishing returns.
    let powerSuitabilityScore = 0;
    if (selectedConn.powerKw >= profile.maxDcKw) {
      // Station delivers car's max ceiling or higher
      const overkill = selectedConn.powerKw - profile.maxDcKw;
      if (overkill > 100) {
        // High overkill (e.g. 240kW for 50kW car)
        powerSuitabilityScore = 20; // good, but slight tariff penalty
      } else {
        // Optimal match
        powerSuitabilityScore = 32;
      }
    } else {
      // Connector slower than car's DC rate
      const ratio = selectedConn.powerKw / profile.maxDcKw;
      powerSuitabilityScore = Math.max(5, Math.round(ratio * 25));
    }

    // 3. Proximity score (closer is much better)
    // 0km -> 40pts, 10km -> 20pts, 20km -> 0pts
    const distanceScore = Math.max(0, 40 - (station.distanceKm * 1.8));

    // 4. Live Availability score
    const availBonus = availableConnectors.length > 0 ? 45 + Math.min(10, availableConnectors.length * 4) : -25;

    // 5. Price factor
    const priceScore = Math.max(0, 30 - selectedConn.pricePerKwh);

    // 6. Rating bonus
    const ratingScore = station.rating * 3; // ~12-15 pts

    const totalScore = Math.round(availBonus + distanceScore + powerSuitabilityScore + priceScore + ratingScore);

    if (totalScore > topScore) {
      topScore = totalScore;
      bestStation = station;
      bestConnector = selectedConn;

      // Formulate clear, insightful explanation
      if (availableConnectors.length > 0 && selectedConn.powerKw >= profile.maxDcKw) {
        matchReason = `Peak ${profile.maxDcKw} kW charging speed match for your ${profile.model} with ${availableConnectors.length} open plug(s).`;
      } else if (availableConnectors.length > 0) {
        matchReason = `Fastest available socket (${selectedConn.powerKw} kW) just ${station.distanceKm} km away.`;
      } else {
        matchReason = `High-power hub with low queue times; slots opening shortly.`;
      }
    }
  }

  if (!bestStation || !bestConnector) return null;

  return {
    station: bestStation,
    score: Math.max(1, Math.min(100, topScore)),
    reason: matchReason,
    recommendedConnector: bestConnector
  };
}

File: app/src/main/java/com/aistudio/chargemate/service/RecommendationEngine.kt
package com.aistudio.chargemate.service

import com.aistudio.chargemate.model.*
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

object RecommendationEngine {
    val EV_PRESETS = listOf(
        EVProfile("nexon-ev", "Tata Motors", "Nexon EV Long Range", 40.5, 50.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 325),
        EVProfile("curvv-ev", "Tata Motors", "Curvv.ev 55", 55.0, 70.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 502),
        EVProfile("punch-ev", "Tata Motors", "Punch.ev Empowered+", 35.0, 50.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 365),
        EVProfile("tiago-ev", "Tata Motors", "Tiago.ev XZ+", 24.0, 25.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 250),
        EVProfile("xuv400", "Mahindra", "XUV400 EL Pro", 39.4, 50.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 375),
        EVProfile("be-6e", "Mahindra", "BE 6e INGLO", 59.0, 175.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 550),
        EVProfile("xev-9e", "Mahindra", "XEV 9e Flagship", 79.0, 175.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 656),
        EVProfile("mg-windsor", "MG Motor", "Windsor EV Essence", 38.0, 45.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 331),
        EVProfile("mg-zs-ev", "MG Motor", "ZS EV Exclusive Plus", 50.3, 75.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 461),
        EVProfile("mg-comet", "MG Motor", "Comet EV Pace", 17.3, 3.3, listOf(ConnectorType.Type2), 230),
        EVProfile("ioniq-5", "Hyundai", "Ioniq 5 AWD (800V)", 72.6, 350.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 631),
        EVProfile("creta-ev", "Hyundai", "Creta Electric", 45.0, 60.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 410),
        EVProfile("kia-ev6", "Kia", "EV6 GT-Line AWD", 77.4, 350.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 708),
        EVProfile("byd-atto3", "BYD", "Atto 3 Blade Dynamic", 60.5, 80.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 521),
        EVProfile("byd-seal", "BYD", "Seal Performance AWD", 82.5, 150.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 580),
        EVProfile("volvo-ex40", "Volvo", "EX40 Recharge Twin", 78.0, 150.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 534),
        EVProfile("bmw-i4", "BMW", "i4 eDrive40 Gran Coupé", 83.9, 205.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 590),
        EVProfile("mercedes-eqb", "Mercedes-Benz", "EQB 350 4MATIC", 70.5, 100.0, listOf(ConnectorType.CCS2, ConnectorType.Type2), 423),
        EVProfile("ather-450x", "Ather Energy", "450X Gen 3 Pro", 3.7, 3.3, listOf(ConnectorType.Type2), 110),
        EVProfile("ola-s1-pro", "Ola Electric", "S1 Pro Gen 2", 4.0, 3.3, listOf(ConnectorType.Type2), 195)
    )

    fun calculateBestMatch(
        stations: List<ChargingStation>,
        profile: EVProfile
    ): RecommendationResult? {
        if (stations.isEmpty()) return null

        var bestStation: ChargingStation? = null
        var topScore = -1000
        var bestConnector: Connector? = null
        var matchReason = ""

        for (station in stations) {
            val compatible = station.connectors.filter { profile.supportedConnectors.contains(it.type) }
            if (compatible.isEmpty()) continue

            val available = compatible.filter { it.status == ConnectorStatus.AVAILABLE }
            val pool = if (available.isNotEmpty()) available else compatible

            val selectedConn = pool.maxByOrNull { it.powerKw } ?: pool.first()

            val powerScore: Int = if (selectedConn.powerKw >= profile.maxDcKw) {
                val overkill = selectedConn.powerKw - profile.maxDcKw
                if (overkill > 100) 20 else 32
            } else {
                val ratio = selectedConn.powerKw / profile.maxDcKw
                max(5, (ratio * 25).roundToInt())
            }

            val distanceScore = max(0.0, 40.0 - (station.distanceKm * 1.8)).roundToInt()
            val availBonus = if (available.isNotEmpty()) 45 + min(10, available.size * 4) else -25
            val priceScore = max(0.0, 30.0 - selectedConn.pricePerKwh).roundToInt()
            val ratingScore = (station.rating * 3.0).roundToInt()

            val totalScore = availBonus + distanceScore + powerScore + priceScore + ratingScore

            if (totalScore > topScore) {
                topScore = totalScore
                bestStation = station
                bestConnector = selectedConn

                matchReason = if (available.isNotEmpty() && selectedConn.powerKw >= profile.maxDcKw) {
                    "Peak ${profile.maxDcKw.roundToInt()} kW speed match for your ${profile.model} with ${available.size} open socket(s)."
                } else if (available.isNotEmpty()) {
                    "Fastest available socket (${selectedConn.powerKw.roundToInt()} kW) just ${station.distanceKm} km away."
                } else {
                    "High-power hub with low queue times; slots opening shortly."
                }
            }
        }

        return if (bestStation != null && bestConnector != null) {
            RecommendationResult(
                station = bestStation,
                score = max(1, min(100, topScore)),
                reason = matchReason,
                recommendedConnector = bestConnector
            )
        } else null
    }
}
